book/website/reproducible-research/reviewing/reviewing-checklist.md

```diff
@@ -1,40 +1,180 @@
+- [ ] Are these modules as small as they can be?
+- [ ] Is there a clear, hierarchical or layered, dependency structure between
+  these modules?
+  - [ ] If not, the functionality should be rearranged, or perhaps heavily
+    interdependent modules should be combined.
+- [ ] Can the design be simplified?
+
+#### Security
+
+If you're making software that is accessible to the outside world (for example a web
+application), then security becomes important. Security issues are defects,
+but not all defects are security issues. A security-conscious design can help
+mitigate the security impact of defects.
+
+- [ ] Which modules deal with user input?
+- [ ] Which modules generate output?
+- [ ] Are input and output compartmentalized?
+  - [ ] If not, consider making separate modules that manage all input
+    and output, so validation can happen in one place.
+- [ ] In which modules is untrusted data present?
+  - [ ] The fewer the better.
+- [ ] Is untrusted data compartmentalized?
+  - [ ] Ideally, validate in the input module and pass only
+    validated data to other parts.
+
+#### Legal
+
+As a developer, you should pay attention to the legal rights of the
+creators of the code you're using. Here are some things to check. When in
+doubt, ask your licensing person for advice.
+
+- [ ] Are the licenses of all modules/libraries that are used documented?
+- [ ] Are the requirements set by those licenses fulfilled?
+  - [ ] Are the licenses included where needed?
+  - [ ] Are copyright statements included in the code where needed?
+  - [ ] Are copyright statements included in the documentation where needed?
+- [ ] Are the licenses of all the parts compatible with each other?
+- [ ] Is the project license compatible with all libraries?
+
+<a name="File/Change level checklist"></a>
+### File/Change level checklist
+
+When you're checking individual changes or files in a pull request, the
+code itself becomes the subject of scrutiny. Depending on the language, files
+may contain interfaces, classes or other type definitions, and functions. All
+these should be checked.
+
+#### Interfaces
+
+- [ ] Is the interface documented?
+- [ ] Does the concept it models make sense?
+- [ ] Can it be split up further? (Interfaces should be as small as possible)
+
+Note that most of the following items assume an object-oriented programming
+style, which may not be relevant to the code you're looking at.
+
+#### Classes and types
+
+- [ ] Is the class documented?
+- [ ] Does it have a single responsibility? Can it be split?
+- [ ] If it's designed to be extended, can it be?
+- [ ] If it's not designed to be extended, is it protected against that?
+- [ ] If it's derived from another class, can you substitute an object of this class for one of its parent class(es)?
+- [ ] Is the class testable?
+  - [ ] Are the dependencies clear and explicit?
+  - [ ] Does it have a small number of dependencies?
+  - [ ] Does it depend on interfaces, rather than on classes?
```

Comment:
External programs needed by classes (for instance command-line tools called from the classes), should also be documented, as they might be optional for several usages. The installation procedure of these external programs should be documented in the installation documentation.

Reply:
In section **Classes and types**, Under the item " Is the class documented", an extra item is added for "External programs needed by classes". Regarding the documentation of the installation procedure of these external programs, I think it is already covered by the item "Is the use and function of third-party libraries documented?" in section **Documentation**. Is it OK?

Reply:
Yes